doc/specs/stdlib_string_type.md

```diff
@@ -0,0 +1,1607 @@
+
+
+## Derived types provided
+
+
+<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
+### The `string_type` derived type
+
+The `string_type` is defined as a non-extenable derived type representing a
+sequence of characters. The internal representation of the character sequence
+is implementation dependent and not visible for the user of the module.
+
+#### Status
+
+Experimental
+
+
+## Procedures and methods provided
+
+Procedures returning `string_type` instances can usually used in elemental
+context, while procedures returning scalar character values can only be
+used in a pure way.
+
+
+<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
+### Constructor for empty string
+
+The module defines a default constructor to create an empty string type.
+
+#### Description
+
+Creates a string instance representing an empty string.
+
+#### Syntax
+
+`res = [[stdlib_string_type(module):string_type(interface)]] ()`
+
+#### Status
+
+Experimental
+
+#### Class
+
+Elemental function.
+
+#### Argument
+
+None.
+
+#### Result value
+
+The result is an instance of `string_type`.
+
+#### Example
+
+```fortran
+use stdlib_string_type
+implicit none
+type(string_type) :: string
+string = string_type()
+! len(string) == 0
+end
+```
+
+
+<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
+### Constructor from character scalar
+
+The module defines a default constructor to create a string type
+from a character scalar.
+
+#### Description
+
+Creates a string instance representing the input character scalar value.
+The constructor shall create an empty string if an unallocated deferred-length
+character variable is passed.
+
+#### Syntax
+
+`res = [[stdlib_string_type(module):string_type(interface)]] (string)`
+
+#### Status
+
+Experimental
+
+#### Class
+
+Elemental function.
+
+#### Argument
+
+`string`: shall be a scalar character value. It is an `intent(in)` argument.
+
+#### Result value
+
+The result is an instance of `string_type`.
+
+#### Example
+
+```fortran
+use stdlib_string_type
+implicit none
+type(string_type) :: string
+string = string_type("Sequence")
+! len(string) == 8
+string = string_type(" S p a c e d ")
+! len(string) == 9
+end
+```
+
+
+<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
+### Assignment of character scalar
+
+The module defines an assignment operations, `=`, to create a string type
+from a character scalar.
+
+#### Description
+
+Creates a string instance representing the right-hand-side character scalar value.
+
+#### Syntax
+
+`lhs = rhs`
+
+#### Status
+
+Experimental
+
+#### Class
+
+Elemntal subroutine, `assignment(=)`.
+
+#### Example
+
+```fortran
+use stdlib_string_type
+implicit none
+type(string_type) :: string
+! len(string) == 0
+string = "Sequence"
+! len(string) == 8
+end
+```
+
+
+<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
+### Len function
+
+#### Description
+
+Returns the length of the character sequence represented by the string.
```

Comment:
Is their a need to carry always the "character sequence" or couldn't this be simply "Returns the length of the string instance"? I am wondering whether we need this distinction.

Reply:
Just some boilerplate I started with and carried on with afterwards. I tried to avoid calling it string because at the point I was not sure if it is going to be *the* string we agree on and looked for something more generic.

Reply:
I think "string" is fine, simpler language.